src/txHashBuilder.c

```diff
@@ -862,9 +887,88 @@ static void txHashBuilder_assertCanLeaveWithdrawals(tx_hash_builder_t* builder)
 	ASSERT(builder->remainingWithdrawals == 0);
 }
 
-void txHashBuilder_addAuxData(tx_hash_builder_t* builder, const uint8_t* auxDataHashBuffer, size_t auxDataHashBufferSize)
+void txHashBuilder_enterMint(tx_hash_builder_t* builder)
 {
 	txHashBuilder_assertCanLeaveWithdrawals(builder);
+	{
+		// Enter mint
+		BUILDER_APPEND_CBOR(CBOR_TYPE_UNSIGNED, TX_BODY_KEY_MINT);
+	}
+	builder->state = TX_HASH_BUILDER_IN_MINT;
+}
+
+void txHashBuilder_addMint_topLevelData(
+        tx_hash_builder_t* builder, uint16_t numAssetGroups
+)
+{
+	_TRACE("state = %u", builder->state);
+
+	ASSERT(builder->state == TX_HASH_BUILDER_IN_MINT);
+
+	builder->outputData.remainingAssetGroups = numAssetGroups;
+	// Map(numAssetGroups)[
+	//   { * policy_id => { * asset_name => uint } }
+	// ]
+	BUILDER_APPEND_CBOR(CBOR_TYPE_MAP, numAssetGroups);
+	if (0 != numAssetGroups) {
+		builder->state = TX_HASH_BUILDER_IN_MINT_ASSET_GROUP;
+	} else {
+		// TODO is this an error case?
```

Comment:
Generally, the tx hash builder is a bit more general than APDU handlers. I.e. it accepts and correctly hashes more than security policies and APDU validation allow. So if this is an error, it would be a VALIDATE error in some APDU handler, not an error here. We can leave a comment like "tx hash builder allows empty mint" --- AFAIK it is not forbidden to have empty mint on the protocol level and it would be accepted by a node, though I'll ask on slack to clarify it.

https://input-output-rnd.slack.com/archives/G010N9UFDNH/p1621934329014800

Please make sure that our implementation is consistent with their answer.

Right now, I have a suspicion that we do allow empty maps, but then the line
```
		ASSERT(subctx->numTokens > 0);
```
looks like we are asserting something that has not been validated before. That would be a bug.